### PR TITLE
Add new notification when a new Node is created in editor

### LINKED ISSUE #7593: Allow scripts to override default property values
### Describe the project you are working on

Maintaining the Godot XR Tools library

### Describe the problem or limitation you are having in your project

Godot XR Tools provides many types of nodes for users to create XR experiences:
image

Godot XR Tools has many scenes which exist purely to provide working default values for properties provided by the C++ scene objects such as collision layers and masks. Our users are continuously trying to use the Add Child Node dialog in the scene-tree editor which results in broken objects due to these properties not being configured correctly.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

We would like a way for the scripts to override the creation values for properties in the base C++ nodes.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

A script can already change the default revert value for C++ implemented properties by implementing the `_property_can_revert()` and `_property_get_revert()` methods.

Changing the initial value after "Add New Node" could be achieved via a new notification (NOTIFICATION_CREATED) which is invoked only when a new node is created (either via Add New Node, or dynamically in code) - not when it is instantiated from a PackedScene or duplicated.

The script would then be able to contain the following code:
```gdscript
const DEFAULT_COLLISION_LAYER = 0x0103

# Override base properties when new node is constructed
func _notification(what : int):
    if what == NOTIFICATION_CREATED:
        collision_layer = DEFAULT_COLLISION_LAYER

# Report revert support for base properties
func _property_can_revert(property : String) -> bool:
    return property == "collision_layer"

# Provide report values for base properties
func _property_get_revert(property : String):
    if property == "collision_layer":
        return DEFAULT_COLLISION_LAYER
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

The current approach of providing pre-configured scenes and extensive user documentation appears to be the only workaround at this time; however it creates significant user issues.

### Is there a reason why this should be core and not an add-on in the asset library?

No means of implementing it outside of the core has been found.

### PR DESCRIPTION
**Pull request by Malcolmnixon**
This pull request implements https://github.com/godotengine/godot-proposals/issues/7593 by:
- Adding a NOTIFICATION_EDITOR_CREATED notification.
- Emitting NOTIFICATION_EDITOR_CREATED when the scene_tree_dock creates new node.

The following script demonstrates how these changes can create a new node script capable of being instantiated via the "Add New Node" scene-tree command and having the node pre-populated with the overridden default values specified by the script.

```gdscript
@tool
class_name CustomRigidBody3D
extends RigidBody3D


# Specify custom default collision layers used in the editor
const DEFAULT_COLLISION_LAYER := 3
const DEFAULT_COLLISION_MASK := 3


# Apply new default values when created in the editor
func _notification(what : int):
	if what == NOTIFICATION_EDITOR_CREATED:
		collision_layer = DEFAULT_COLLISION_LAYER
		collision_mask = DEFAULT_COLLISION_MASK


# Indicate revert supported in the editor inspector
func _property_can_revert(property : StringName) -> bool:
	return property == "collision_layer" or property == "collision_mask"


# Provide revert values to the editor inspector
func _property_get_revert(property : StringName):
	match property:
		"collision_layer":
			return DEFAULT_COLLISION_LAYER

		"collision_mask":
			return DEFAULT_COLLISION_MASK
```

*Bugsquad edit:* Closes https://github.com/godotengine/godot-proposals/issues/7593

### CODE REVIEW

**Comment by MewPurPur** [comment]
Seems like I have been doing this in my game like

```
func _enter_tree() -> void:
	if SceneUtils.is_tool():
		z_index = 6
		z_as_relative = false
```

I suppose this made it so these defaults are enforced every single time the scene is opened, while your PR adds a way to enforce a default only when the node is added?

**Comment by YuriSizov** [comment]
I'm concerned about adding an editor-specific notification to Node, but this is likely useful and we do seem to have a couple other editor notifications in there.

You probably need to add this same notification to `SceneCreateDialog::create_scene_root()` and `SceneTreeDock::_tool_selected()` (for `TOOL_CREATE_*` options).

**Comment by AThousandShips** [comment]
Think the documentation should clarify that it happens when created through the UI, and not through for example a script

**Comment by Malcolmnixon** [comment]
I had considered other methods for allowing scripts to override default values; however one of my primary goals was to limit the performance impact on shipping games. The three approaches I considered were:

- Forcing the editor to "revert" all properties on new nodes:
   - Allows changing default values using only `_property_can_revert()` and `_property_get_revert()`
   - May interfere with values set during `_init()`
- Adding new `@override` attribute:
   - Significant changes to scripting
   - Might introduce ambiguities with `_property_can_revert()` and `_property_get_revert()`
   - May incur runtime cost of processing override list in object construction
- Adding the NOTIFICATION_EDITOR_CREATED signal:
   - Virtually no performance impact on shipping games
   - May be useful for other use cases
   - Does not support nodes created in scripts, however:
       - The script creating the node can set properties
       - The script creating the node could choose to send the notification

**Comment by BastiaanOlij** [comment]
cc @vnen 

What is your view on this George? Is this the proper way with this, it's something we need to continue some of the development on XR Tools for.

**Comment by KoBeWi** [comment]
If it does solve the problem, I think this implementation is fine.
There is one more way of creating nodes that I think might need to receive the notification - when you drop a script to scene tree while holding Ctrl. It creates a new instance of the type and it's a UI action, so I think it should be handled too.

**Comment by Malcolmnixon** [comment]
> If it does solve the problem, I think this implementation is fine. There is one more way of creating nodes that I think might need to receive the notification - when you drop a script to scene tree while holding Ctrl. It creates a new instance of the type and it's a UI action, so I think it should be handled too.

I added firing NOTIFICATION_EDITOR_CREATED for the Ctrl script-drag into the scene tree.

**Review by KoBeWi — DISMISSED** [review]

**Comment by akien-mga** [comment]
Asked reduz to review, but in typical reduz fashion he replied to me instead of commenting here ;)

```
<reduz> I dont understand, they want a new notification `NOTIFICATION_CREATED` basically? you can do this already `if (Engine.is_editor_hint() and get_tree().get_edited_scene_root().is_ancestor_of(this)`)
<Akien> Sounds like it yeah.
<Akien> But I guess they didn't know they could do it like this.
<YuriSizov> They seem to need it outside of tree.
```

**Comment by KoBeWi** [comment]
There isn't any `NOTIFICATION_CREATED` and even if it was, it's not the same as the new one here. It needs to be received only when the node is created *for the first time*.

**Comment by KoBeWi** [comment]
> Notification received when the object is initialized, before its script is attached. Used internally.

Not really. And as I said, this notification is received when node is duplicated, instantiated etc. It's not suitable for initializing a default value, because it would be overwriting the property every time.

**Comment by BastiaanOlij** [comment]
Just to bring this back to attention. Discussing this during the XR meeting where we are waiting for this functionality.

Let's focus back on the original problem we are trying to solve here. Right now it is not possible to extend a class and provide alternative default values for properties from the class we are extending.

So RigidBody3D has a default collision layer of 1, in our extended class we want that default layer to be 3.

We can do this when we also defined a `custom_rigid_body_3d.tscn` along side the new class and specify the overriding default there, however when using "add new node" instead of "add child scene", this accompanying scene file is not used.

The bottom line here is solving that issue, this PR solves that by introducing a notification after our node is created, but before our properties are processed, so that the default value can be overridden. The other suggestions made here as alternatives do not fit that purpose, they either happen to early, too late, or too often in order to solve the underlying issue.

**Comment by Calinou** [comment]
Regarding use cases, check the bottom of the description of https://github.com/godotengine/godot/pull/88647. I tried to implement a virtual method for this, but the method is never being called: https://github.com/Calinou/godot/tree/editor-add-on-instantiated-call

**Comment by KoBeWi** [comment]
I'm getting reminded about this PR once every few months, because new scenarios keep popping up where such notification would be useful.

> I would prefer we find another way to do this than hacking a notification inside Node.

This change does not modify node in any way, other than registering a new constant. The notification could be registered elsewhere, but being next to _other editor-specific notifications_ is more intuitive I think.

> As I mentioned above, there are other ways to work around this issue.

None of the existing methods/notifications allow to achieve the same thing (unless I'm missing something), so what are these other ways?

Again, the goal is to notify the node when it's just created in the editor, once. It's similar to `PROPERTY_USAGE_EDITOR_INSTANTIATE_OBJECT`, but for nodes.

File: editor/scene_create_dialog.cpp
Comment (KoBeWi) [inline editor/scene_create_dialog.cpp:173]: I guess it might be appropriate to move all calls to `instantiate_object_properties()` instead, it's called in the same scenarios.

**Comment by KoBeWi** [comment]
Superseded by #122058
Thanks for the contribution.
